File: Projetos/ACBrLib/Demos/Java/Mail/Imports/ACBrLibMail/src/com/acbr/mail/ACBrMail.java
package com.acbr.mail;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.ptr.IntByReference;
import java.io.File;
import com.acbr.ACBrSessao;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.file.Paths;

public final class ACBrMail  {    
       
    private interface ACBrMailLib extends Library {
        
        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrMailLib INSTANCE = LibraryLoader.getInstance();
        
        class LibraryLoader {
            private static String library = "";
            private static ACBrMailLib instance = null;
            
            public static String getLibraryName() {
                if (library.isEmpty()) {
                    library = Platform.is64Bit() ? "ACBrMail64" : "ACBrMail32";
                }
                
                return library;
            }
            
            public static ACBrMailLib getInstance() {
                if (instance == null) {
                    instance = (ACBrMailLib) Native.synchronizedLibrary((Library) Native.loadLibrary(JNA_LIBRARY_NAME, ACBrMailLib.class));
                }
                
                return instance;
            }
        }
        
        int MAIL_Inicializar(String eArqConfig, String eChaveCrypt);
        
        int MAIL_Finalizar();
        
        int MAIL_Nome(ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_Versao(ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_UltimoRetorno(ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_ConfigLer(String eArqConfig);
        
        int MAIL_ConfigGravar(String eArqConfig);
        
        int MAIL_ConfigLerValor(String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_ConfigGravarValor(String eSessao, String eChave, String valor);
          
        int MAIL_SetSubject(String eSubject);  
        
        int MAIL_AddAddress(String eEmail, String eName);
        
        int MAIL_AddReplyTo(String eEmail, String eName);
        
        int MAIL_AddCC(String eEmail, String eName);
        
        int MAIL_AddBCC(String eEmail);
        
        int MAIL_ClearAttachment();
        
        int MAIL_AddAttachment(String eFileName, String eDescription, int aDisposition);
        
        int MAIL_AddBody(String eBody);
        
        int MAIL_AddAltBody(String eAltBody);
        
        int MAIL_SaveToFile(String eFileName);
        
        int MAIL_Clear();
        
        int MAIL_Send();
    }

    private static Charset UTF8 = Charset.forName("UTF-8");
    private static final int STR_BUFFER_LEN = 256;
    
    public ACBrMail() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }

        int ret = ACBrMailLib.INSTANCE.MAIL_Inicializar(toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
    }

    public ACBrMail(String eArqConfig, String eChaveCrypt) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_Inicializar(toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
    }
    
    @Override
    protected void finalize() throws Throwable {
        try {
            int ret = ACBrMailLib.INSTANCE.MAIL_Finalizar();
            checkResult(ret);
        } finally {
            super.finalize();
        }
    }
    
    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_Nome(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_Versao(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigLer(toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigGravar(toUTF8(eArqConfig));
        checkResult(ret);
    }

    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigLerValor(toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigGravarValor(toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }
    
    public void setSubject(String eMail) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_SetSubject(toUTF8(eMail));
        checkResult(ret);
    }
    
    public void addAddress(String eMail, String eName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddAddress(toUTF8(eMail), toUTF8(eName));
        checkResult(ret);
    }
    
    public void addReplyTo(String eMail, String eName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddReplyTo(toUTF8(eMail), toUTF8(eName));
        checkResult(ret);
    }
    
    public void addCC(String eMail, String eName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddCC(toUTF8(eMail), toUTF8(eName));
        checkResult(ret);
    }
    
    public void addBCC(String eMail) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddBCC(toUTF8(eMail));
        checkResult(ret);
    }
    
    public void clearAttachment() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ClearAttachment();
        checkResult(ret);
    }
    
    public void addAttachment(String eFileName, String eDescription, MailAttachmentDisposition aDisposition) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddAttachment(toUTF8(eFileName), 
                toUTF8(eDescription), aDisposition.asInt());
        checkResult(ret);
    }
    
    public void addBody(String eBody) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddBody(toUTF8(eBody));
        checkResult(ret);
    }
    
    public void addAltBody(String eAltBody) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddAltBody(toUTF8(eAltBody));
        checkResult(ret);
    }
    
    public void saveToFile(String eFileName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_SaveToFile(toUTF8(eFileName));
        checkResult(ret);
    }
    
    public void clear() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_Clear();
        checkResult(ret);
    }
    
    public void send() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_Send();
        checkResult(ret);
    }
    
    private static String toUTF8(String value) {
        return new String(value.getBytes(UTF8));
    }

    private static String fromUTF8(ByteBuffer buffer, int len) {
        return new String(buffer.array(), 0, len, UTF8);
    }

    private static void checkResult(int result) throws Exception {
        if (result == 0) {
            return;
        }

        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        ACBrMailLib.INSTANCE.MAIL_UltimoRetorno(buffer, bufferLen);
        throw new Exception(processResult(buffer, bufferLen));
    }
    
    private static String processResult(ByteBuffer buffer, IntByReference bufferLen){
        if (bufferLen.getValue() <= STR_BUFFER_LEN) {
            return fromUTF8(buffer, bufferLen.getValue());
        }

        if (bufferLen.getValue() > STR_BUFFER_LEN) {
            buffer = ByteBuffer.allocate(bufferLen.getValue());
            ACBrMailLib.INSTANCE.MAIL_UltimoRetorno(buffer, bufferLen);
        }

        return fromUTF8(buffer, bufferLen.getValue());
    }
}

File: Projetos/ACBrLib/Demos/C#/Mail/Imports/Dinamico/ACBrMail.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using ACBrLib.Core;
using ACBrLib.Core.Mail;

namespace ACBrLib.Mail
{
    public sealed class ACBrMail : ACBrLibHandle
    {
        #region InnerTypes

        private class Delegates
        {
            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddAddress(string eEmail, string eName);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddAltBody(string eAltBody);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddAttachment(string eFileName, string eDescription, int aDisposition);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddBCC(string eEmail);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddBody(string eBody);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddCC(string eEmail, string eName);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_AddReplyTo(string eEmail, string eName);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_Clear();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_ClearAttachment();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_ConfigGravar(string eArqConfig);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_ConfigGravarValor(string eSessao, string eChave, string valor);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_ConfigLer(string eArqConfig);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_ConfigLerValor(string eSessao, string eChave, StringBuilder buffer,
                ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_Finalizar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_Inicializar(string eArqConfig, string eChaveCrypt);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_Nome(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_SaveToFile(string eFileName);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_Send();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_SetSubject(string eSubject);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_UltimoRetorno(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int MAIL_Versao(StringBuilder buffer, ref int bufferSize);
        }

        #endregion InnerTypes

        #region Constructors

        public ACBrMail(string eArqConfig = "", string eChaveCrypt = "") :
            base(Environment.Is64BitProcess ? "ACBrMail64.dll" : "ACBrMail32.dll")
        {
            InitializeMethods();

            var inicializar = GetMethod<Delegates.MAIL_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);
        }

        #endregion Constructors

        #region Methods

        #region Ini

        public void ConfigGravar(string eArqConfig = "ACBrLib.ini")
        {
            var gravarIni = GetMethod<Delegates.MAIL_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public void ConfigLer(string eArqConfig = "ACBrLib.ini")
        {
            var lerIni = GetMethod<Delegates.MAIL_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<Delegates.MAIL_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = FromUTF8(pValue);

            if (typeof(T).IsEnum) return (T)Enum.ToObject(typeof(T), Convert.ToInt32(value));

            if (typeof(T) == typeof(bool)) return (T)(object)Convert.ToBoolean(Convert.ToInt32(value));

            return (T)Convert.ChangeType(value, typeof(T));
        }

        public void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<Delegates.MAIL_ConfigGravarValor>();
            var type = value.GetType();

            var propValue = value.ToString();
            if (type.IsEnum) propValue = ((int)value).ToString();
            if (type == typeof(bool)) propValue = Convert.ToInt32(value).ToString();

            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        #endregion Ini

        public void SetSubject(string subject)
        {
            var method = GetMethod<Delegates.MAIL_SetSubject>();
            var ret = ExecuteMethod(() => method(ToUTF8(subject)));

            CheckResult(ret);
        }

        public void AddAddress(string eEmail, string eName)
        {
            var method = GetMethod<Delegates.MAIL_AddAddress>();
            var ret = ExecuteMethod(() => method(ToUTF8(eEmail), ToUTF8(eName)));

            CheckResult(ret);
        }

        public void AddReplyTo(string eEmail, string eName)
        {
            var method = GetMethod<Delegates.MAIL_AddReplyTo>();
            var ret = ExecuteMethod(() => method(ToUTF8(eEmail), ToUTF8(eName)));

            CheckResult(ret);
        }

        public void AddCC(string eEmail, string eName)
        {
            var method = GetMethod<Delegates.MAIL_AddCC>();
            var ret = ExecuteMethod(() => method(ToUTF8(eEmail), ToUTF8(eName)));

            CheckResult(ret);
        }

        public void AddBCC(string eEmail, string eName)
        {
            var method = GetMethod<Delegates.MAIL_AddBCC>();
            var ret = ExecuteMethod(() => method(ToUTF8(eEmail)));

            CheckResult(ret);
        }

        public void AddAttachment(string eFileName, string eDescription, MailAttachmentDisposition aDisposition)
        {
            var method = GetMethod<Delegates.MAIL_AddAttachment>();
            var ret = ExecuteMethod(() => method(ToUTF8(eFileName), ToUTF8(eDescription), (int)aDisposition));

            CheckResult(ret);
        }

        public void ClearAttachment()
        {
            var method = GetMethod<Delegates.MAIL_ClearAttachment>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void AddBody(string eBody)
        {
            var method = GetMethod<Delegates.MAIL_AddBody>();
            var ret = ExecuteMethod(() => method(ToUTF8(eBody)));

            CheckResult(ret);
        }

        public void AddAltBody(string eAltBody)
        {
            var method = GetMethod<Delegates.MAIL_AddAltBody>();
            var ret = ExecuteMethod(() => method(ToUTF8(eAltBody)));

            CheckResult(ret);
        }

        public void SaveToFile(string eFileName)
        {
            var method = GetMethod<Delegates.MAIL_SaveToFile>();
            var ret = ExecuteMethod(() => method(ToUTF8(eFileName)));

            CheckResult(ret);
        }

        public void Clear()
        {
            var method = GetMethod<Delegates.MAIL_Clear>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void Send()
        {
            var method = GetMethod<Delegates.MAIL_Send>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<Delegates.MAIL_Finalizar>();
            var codRet = ExecuteMethod(() => finalizar());
            CheckResult(codRet);
        }

        protected override string GetUltimoRetorno()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<Delegates.MAIL_UltimoRetorno>();

            ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));

            if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

            buffer.Capacity = bufferLen;
            ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));

            return FromUTF8(buffer);
        }

        private void InitializeMethods()
        {
            AddMethod<Delegates.MAIL_Inicializar>("MAIL_Inicializar");
            AddMethod<Delegates.MAIL_Finalizar>("MAIL_Finalizar");
            AddMethod<Delegates.MAIL_Nome>("MAIL_Nome");
            AddMethod<Delegates.MAIL_Versao>("MAIL_Versao");
            AddMethod<Delegates.MAIL_UltimoRetorno>("MAIL_UltimoRetorno");
            AddMethod<Delegates.MAIL_ConfigLer>("MAIL_ConfigLer");
            AddMethod<Delegates.MAIL_ConfigGravar>("MAIL_ConfigGravar");
            AddMethod<Delegates.MAIL_ConfigLerValor>("MAIL_ConfigLerValor");
            AddMethod<Delegates.MAIL_ConfigGravarValor>("MAIL_ConfigGravarValor");
            AddMethod<Delegates.MAIL_SetSubject>("MAIL_SetSubject");
            AddMethod<Delegates.MAIL_AddAddress>("MAIL_AddAddress");
            AddMethod<Delegates.MAIL_AddReplyTo>("MAIL_AddReplyTo");
            AddMethod<Delegates.MAIL_AddCC>("MAIL_AddCC");
            AddMethod<Delegates.MAIL_AddBCC>("MAIL_AddBCC");
            AddMethod<Delegates.MAIL_AddAttachment>("MAIL_AddAttachment");
            AddMethod<Delegates.MAIL_ClearAttachment>("MAIL_ClearAttachment");
            AddMethod<Delegates.MAIL_AddBody>("MAIL_AddBody");
            AddMethod<Delegates.MAIL_AddAltBody>("MAIL_AddAltBody");
            AddMethod<Delegates.MAIL_SaveToFile>("MAIL_SaveToFile");
            AddMethod<Delegates.MAIL_Clear>("MAIL_Clear");
            AddMethod<Delegates.MAIL_Send>("MAIL_Send");
        }

        #endregion Private Methods

        #endregion Methods
    }
}